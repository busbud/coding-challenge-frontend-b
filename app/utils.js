// UTC time (dateTtime) -> hh:mm
export const utcTimeToHhMm = utcTime => utcTime.split('T')[1].substr(0, 5)

// Number XXXX -> XX,XX (currency)
export const numToCurrency = num => (num/100).toFixed(2)
