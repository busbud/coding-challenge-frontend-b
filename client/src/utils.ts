export const toCamel = (s: string) => {
  return s.replace(/([-_][a-z])/gi, ($1) => {
    return $1.toUpperCase().replace("-", "").replace("_", "");
  });
};

export const capitalize = (str: string) => {
  const lower = str.toLowerCase();
  return str.charAt(0).toUpperCase() + lower.slice(1);
};
