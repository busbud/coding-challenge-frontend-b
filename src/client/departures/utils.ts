import { LuxonDateTime } from '../dependencies';
import { City, DepartureResponse, Location, Price } from '../../api';

const invalidLocation: Location = {
  id: Math.random(),
  name: 'Invalid',
  arrival_timezone: '',
  departure_timezone: '',
  city_id: Math.random().toString(16),
};

const invalidCity: City = {
  id: Math.random().toString(16),
  name: 'Invalid',
  timezone: '',
};

export const getCity = (cities: City[], cityId: string) =>
  cities.find((c) => c.id === cityId) ?? invalidCity;

export const getLocation = (locations: Location[], locationId: number) =>
  locations.find((l) => l.id === locationId) ?? invalidLocation;

export const enrichDepartures = (response: DepartureResponse) => {
  const {
    origin_city_id,
    destination_city_id,
    cities,
    locations,
    departures,
  } = response;

  return departures.map((departure) => {
    const { destination_location_id } = departure;
    const destination_location = getLocation(
      locations,
      destination_location_id,
    );

    return {
      ...departure,
      destination_location,
    };
  });
};

export const getPrice = ({ total, currency }: Price) => `${total} ${currency}`;

export const timeToLocalTime = (time: string, zone: string) =>
  LuxonDateTime.fromISO(time, { zone }).toLocaleString(
    LuxonDateTime.DATETIME_FULL,
  );
