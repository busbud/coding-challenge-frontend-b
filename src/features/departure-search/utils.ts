export function formatTimeToLocale(
  date: Date | string,
  locale: string = 'en'
): string {
  const newDate = new Date(date)
  return newDate.toLocaleTimeString(locale, {
    hour: '2-digit',
    minute: '2-digit',
  })
}

export function formatCurrencyToLocale(
  price: number,
  currency: string,
  locale: string = 'en'
): string {
  return new Intl.NumberFormat(locale, {
    style: 'currency',
    currency,
  }).format(price)
}
