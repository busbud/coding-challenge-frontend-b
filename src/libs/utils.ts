import type { DeparturesResponse, Search } from "../types";

export function getBaseQuery(parameters: Search) {
  return {
    url: `/x-departures/${parameters.origin}/${parameters.destination}/${parameters.date}`,
    params: {
      adult: parameters.passengers,
    },
  };
}

export function getPollQuery(parameters: Search & { index: number }) {
  const baseQuery = getBaseQuery(parameters);
  return {
    url: `${baseQuery.url}/poll`,
    params: {
      adult: parameters.passengers,
      index: parameters.index,
    },
  };
}

export function getLocationNamesById(
  locations: DeparturesResponse["locations"] = []
): { [key: number]: string } {
  return locations.reduce(
    (acc, location) => ({ ...acc, [location.id]: location.name }),
    {}
  );
}
