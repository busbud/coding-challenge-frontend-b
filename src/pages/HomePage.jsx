import React, { useState } from "react";
import { Header } from "../components/header/Header";
import { SearchBar } from "../components/searchBar/SearchBar";
import { TripList } from "../components/tripList/TripList";
import styled, { keyframes } from "styled-components";
import background from "../assets/background.jpg";
import { useTranslation } from "react-i18next";
import search from "../assets/search.svg";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  background: #f7f7f7;
  height: 100%;
`;

const TopSectionContainer = styled.div`
    background-image: url(${background});
    background-image: linear-gradient(to bottom, rgba(0,0,0,0.3) 0%,rgba(0,0,0,0.6) 100%), url(${background});
    background-size: cover;
    background-repeat: no-repeat;
    background-position-y: bottom;
    display: flex;
    width: 100%
    min-height: 50vh;
    align-items: center;
    justify-content: center;
    flex-direction: column
`;

const Title = styled.span`
  color: white;
  font-size: 32px;
  font-weight: bold;
`;

const BounceAnimation = keyframes`
  0% { margin-bottom: 0; }
  50% { margin-bottom: 15px }
  100% { margin-bottom: 0 }
`;

const DotWrapper = styled.div`
  display: flex;
  padding: 2vh 2vw;
  align-items: flex-end;
  height: 100%;
  align-items: center;
`;

const Dot = styled.div`
  background: #f19020;
  border-radius: 50%;
  width: 10px;
  height: 10px;
  margin: 0 5px; /* Animation */
  animation: ${BounceAnimation} 0.5s linear infinite;
  animation-delay: ${props => props.delay};
  box-shadow: 2px 5px 4px #d5d5d5;
`;

const Tips = styled.div`
  white-space: nowrap;
  padding: 2vh 2vw;
  align-items: flex-end;
  height: 100%;
  align-items: center;
  font-size: 13px;
  color: #696969;
`;

const Icon = styled.img`
  max-width: 15px;
  max-height: 15px;
  padding: 0 5px;
`;

export const HomePage = () => {
  const [trips, setTrips] = useState(null);
  const { t, _ } = useTranslation();
  return (
    <>
      <Header />
      <Container>
        <TopSectionContainer>
          <Title> {t("title")}</Title>
          <SearchBar setTrips={setTrips} />
        </TopSectionContainer>
        {trips &&
          (trips.length > 0 ? (
            <TripList trips={trips} />
          ) : (
            <DotWrapper>
              <Dot delay="0s" />
              <Dot delay=".1s" />
              <Dot delay=".2s" />
            </DotWrapper>
          ))}
        {!trips && (
          <Tips>
            {t("tips")} <Icon src={search} />
          </Tips>
        )}
      </Container>
    </>
  );
};
