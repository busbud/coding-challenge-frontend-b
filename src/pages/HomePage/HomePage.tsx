import React from "react";
import { Link } from "react-router-dom";

function HomePage(): JSX.Element {
  return (
    <div className="max-w-lg mx-auto text-center">
      <h1 className="text-3xl font-bold text-gray-800 dark:text-white md:text-4xl">
        Find a ticket for our next Osheaga Musique et art Festival
      </h1>
      <p className="my-6 text-gray-500 dark:text-gray-300">
        This event will take place on 20 December 2021
      </p>
      <Link
        to="/search"
        className="px-6 py-2 text-sm font-medium leading-5 text-center text-white capitalize bg-blue-600 rounded-lg hover:bg-blue-500 md:mx-0 md:w-auto focus:outline-none"
      >
        Find your ticket
      </Link>
    </div>
  );
}

export default HomePage;
