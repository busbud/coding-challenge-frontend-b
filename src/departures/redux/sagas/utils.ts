/* eslint-disable @typescript-eslint/no-explicit-any */
import type { IDeparture } from 'src/types';
import { dateCompare } from 'src/utils/compare';
import { getTime } from 'src/utils/date';

export const createDepartures = (
    responseData: Record<string, any>,
): IDeparture[] => {
    const departures = responseData.departures.map((data: any) => {
        return {
            key: data.id,
            id: data.id,
            departureTime: getTime(data.departure_time),
            arrivalTime: getTime(data.arrival_time),
            price: data.prices.total,
            location: responseData.locations?.find(
                (x: any) => x.id == data.origin_location_id,
            ).name,
            stops: data.trip_stops.map((stop: any) => {
                return {
                    departureTime: getTime(stop.departure_time),
                    arrivalTime: getTime(stop.arrival_time),
                    location: stop.name,
                    duration: stop.duration,
                };
            }),
        } as IDeparture;
    });

    return departures.sort((a: IDeparture, b: IDeparture) =>
        dateCompare(a.departureTime, b.departureTime),
    );
};
