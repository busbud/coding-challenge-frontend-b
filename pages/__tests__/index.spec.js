import * as React from "react";

import HomePage from "../index";
import { setupComponent } from "../../utils/jest";

describe("Home Page", () => {
  it("should render without throwing an error", function() {
    const { asFragment } = setupComponent(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
