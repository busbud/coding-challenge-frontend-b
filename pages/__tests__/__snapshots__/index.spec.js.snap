// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Index should render without throwing an error 1`] = `ShallowWrapper {}`;
